addons/product/models/product_pricelist.py

```diff
@@ -305,33 +305,42 @@ def _get_partner_pricelist(self, partner_id, company_id=None):
              instead of current user's company
         """
-        Partner = self.env['res.partner']
-        Property = self.env['ir.property'].with_context(force_company=company_id or self.env.user.company_id.id)
-
-        p = Partner.browse(partner_id)
-        pl = Property.get('property_product_pricelist', Partner._name, '%s,%s' % (Partner._name, p.id))
-        if pl:
-            pl = pl[0].id
-
-        if not pl:
-            if p.country_id.code:
-                pls = self.env['product.pricelist'].search([('country_group_ids.country_ids.code', '=', p.country_id.code)], limit=1)
-                pl = pls and pls[0].id
-
-        if not pl:
-            # search pl where no country
-            pls = self.env['product.pricelist'].search([('country_group_ids', '=', False)], limit=1)
-            pl = pls and pls[0].id
+        res = self._get_partner_pricelist_multi([partner_id], company_id)
+        return res[partner_id].id
 
-        if not pl:
-            prop = Property.get('property_product_pricelist', 'res.partner')
-            pl = prop and prop[0].id
+    def _get_partner_pricelist_multi(self, partner_ids, company_id=None):
+        """ Retrieve the applicable pricelist for given partners in a given company.
 
-        if not pl:
-            pls = self.env['product.pricelist'].search([], limit=1)
-            pl = pls and pls[0].id
+            :param company_id: if passed, used for looking up properties,
+                instead of current user's company
+            :return: a dict {partner_id: pricelist}
+        """
+        Partner = self.env['res.partner']
+        Property = self.env['ir.property'].with_context(force_company=company_id or self.env.user.company_id.id)
 
-        return pl
+        # retrieve values of property
+        result = Property.get_multi('property_product_pricelist', Partner._name, partner_ids)
+
+        # group partners without value, and find a pricelist for each group
+        remaining_partner_ids = [pid for pid, val in result.items() if not val]
+        if remaining_partner_ids:
+            # Get fallback values
+            pl_fallback = (
+                self.env['product.pricelist'].search([('country_group_ids', '=', False)], limit=1) or
+                Property.get('property_product_pricelist', 'res.partner') or
+                self.env['product.pricelist'].search([], limit=1)
+            )
+
+            # Group by country
+            domain = [('id', 'in', remaining_partner_ids)]
+            groups = Partner.read_group(domain, ['country_id'], ['country_id'])
+            for group in groups:
+                country_id = group['country_id'] and group['country_id'][0]
+                pl_country_group = self.env['product.pricelist'].search([('country_group_ids.country_ids', '=', country_id)], limit=1)
+                pl = pl_country_group or pl_fallback
+                for pid in Partner.search(group['__domain']).ids:
```

Comment:
In case of a long list of remaining partners (let's say 1M), is this efficient?

Maybe this?
```python
for pid in (Partner.search([('country_id', '=', country_id)]).ids) & set(remaining_partner_ids):
....
```

Or could be overkill or useless, I have no idea (and no benchmark in mind).

Reply:
Because `read_group` is invoked with domain
```python
[('id', 'in', remaining_partner_ids)]
```
the domain `group['__domain']` is something like
```python
['&', ('id', 'in', remaining_partner_ids), ('country_id', '=', country_id)]
```
So the query is already the most specific you can get.